src/sagemaker/workflow/pipeline.py

```diff
@@ -534,3 +535,126 @@ def wait(self, delay=30, max_attempts=60):
             waiter_id, model, self.sagemaker_session.sagemaker_client
         )
         waiter.wait(PipelineExecutionArn=self.arn)
+
+
+class PipelineGraph:
+    """Helper class representing the Pipeline Directed Acyclic Graph (DAG)
+
+    Attributes:
+        steps (Sequence[Union[Step, StepCollection]]): Sequence of `Step`s and/or `StepCollection`s
+            that represent each node in the pipeline DAG
+    """
+
+    def __init__(self, steps: Sequence[Union[Step, StepCollection]]):
+        self.step_map = PipelineGraph._generate_step_map(steps)
+        self.adjacency_list = self._initialize_adjacency_list()
+        if self.is_cyclic():
+            raise ValueError("Cycle detected in pipeline step graph.")
+
+    @staticmethod
+    def _generate_step_map(
+        steps: Sequence[Union[Step, StepCollection]]
+    ) -> Dict[str, Union[Step, StepCollection]]:
+        """Helper method to create a mapping from Step/Step Collection name to itself."""
+        step_map = {}
+        for step in steps:
+            if isinstance(step, Step):
+                if step.name in step_map:
+                    raise ValueError("Pipeline steps cannot have duplicate names.")
+                step_map[step.name] = step
+                if isinstance(step, ConditionStep):
+                    step_map.update(
+                        PipelineGraph._generate_step_map(step.if_steps + step.else_steps)
+                    )
```

Comment:
The names of steps in if/else branches can't conflict with steps outside. So it is better to do something like 

```
for step in steps:
    if step.name in self.step_map:
        raise ValueError
    self.step_map[step.name] = step
    if isinstance(step, ConditionStep):
        self.generate_step_map(...)
    elif isinstance(step, StepCollection):
        self.generate_step_map(step.steps) 
```